### PR TITLE
[GPU] Range v4 partial implementation

### PR DESCRIPTION
**Pull request by serhii-pavlovskyi-altran**
unit test works now, single layer needs dynamic shapes

### CODE REVIEW

File: src/plugins/intel_gpu/src/plugin/ops/range.cpp [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+                          const std::shared_ptr<ngraph::op::v4::Range>& op) {
+  p.ValidateInputs(op, {3});
+  auto& outShape = op->get_output_partial_shape(0);
+  auto outRank = outShape.rank();
+  {
+    auto r = outRank.get_length();
+    if (r != 1)
+      throw std::runtime_error{"range v4 output rank is " + std::to_string(r)};
+  }
+  cldnn::tensor outTensor {tensor_from_dims(op->get_output_shape(0))};
```
Comment (sshlyapn) [inline src/plugins/intel_gpu/src/plugin/ops/range.cpp:25]: This will create such tensor in case of `rank== 1`: `cldnn::tensor(cldnn::batch(dims[0]), cldnn::feature(def), cldnn::spatial(def, def)) `
But your kernel expects data in other dimensions (`zyx`) - this behavior should be aligned

Comment (serhii-pavlovskyi-altran) [reply src/plugins/intel_gpu/src/plugin/ops/range.cpp:25]: i think kernel expects larger rank because there's no way to have rank 1 in cldnn. semantically range output is rank 1, i.e. all additional dimensions have length 1

File: inference-engine/thirdparty/clDNN/kernel_selector/core/cl_kernels/range_ref.cl [hunk 2]
```diff
@@ -0,0 +1,11 @@
+// Copyright (C) 2021 Intel Corporation
+// SPDX-License-Identifier: Apache-2.0
+//
+
+KERNEL (range_ref)(const __global INPUT0_TYPE* startP, const __global INPUT1_TYPE* stopP, const __global INPUT2_TYPE* stepP,
+		__global OUTPUT_TYPE *output)
+
+{
+	const uint i = get_global_id(2);
+	output[i] = *startP + i * *stepP;
```
[the remark this answers is not in the document]
Comment (serhii-pavlovskyi-altran) [reply inference-engine/thirdparty/clDNN/kernel_selector/core/cl_kernels/range_ref.cl:10]: can i just use same type as output type(same should pass "better or equal" check)?
i.e.
```
output[i] = *startP + i * (OUTPUT_TYPE)*stepP;
```
promotions should take care of other casts

File: inference-engine/thirdparty/clDNN/kernel_selector/core/actual_kernels/range/range_kernel_ref.cpp [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+  auto &start = params.inputs[0];
+  if (start.LogicalSize() != 1)
+    return false;
+  auto &stop = params.inputs[1];
+  if (stop.LogicalSize() != 1)
+    return false;
+  auto &step = params.inputs[2];
+  if (step.LogicalSize() != 1)
+    return false;
+//  if (params.output.Dimentions() != 1) return false;
```
[the remark this answers is not in the document]
Comment (serhii-pavlovskyi-altran) [reply inference-engine/thirdparty/clDNN/kernel_selector/core/actual_kernels/range/range_kernel_ref.cpp]: done and simplified this function

File: inference-engine/thirdparty/clDNN/kernel_selector/core/actual_kernels/range/range_kernel_ref.cpp [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+
+    FillCLKernelData(kernel_data.kernels[0], dispatch_data, params.engineInfo,
+            kernelName, jit, entry_point, DEFAULT,
+            false, false, 3);
+
+    return kernels_data;
+}
+
+KernelsPriority RangeKernelRef::GetKernelsPriority(const Params& /*params*/, const optional_params& /*options*/) const {
+    return FORCE_PRIORITY_1;
```
[the remark this answers is not in the document]
Comment (serhii-pavlovskyi-altran) [reply inference-engine/thirdparty/clDNN/kernel_selector/core/actual_kernels/range/range_kernel_ref.cpp]: done

File: inference-engine/thirdparty/clDNN/kernel_selector/core/actual_kernels/range/range_kernel_ref.cpp [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+  dispatchData.lws =
+      GetOptimalLocalWorkGroupSizes(dispatchData.gws, params.engineInfo);
+
+  return dispatchData;
+}
+
+}  // namespace
+
+KernelsData RangeKernelRef::GetKernelsData(const Params &params,
+        const optional_params &options) const {
```
[the remark this answers is not in the document]
Comment (serhii-pavlovskyi-altran) [reply inference-engine/thirdparty/clDNN/kernel_selector/core/actual_kernels/range/range_kernel_ref.cpp]: reformatted to one line

File: inference-engine/thirdparty/clDNN/kernel_selector/core/actual_kernels/range/range_kernel_ref.cpp [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+#include "range_kernel_ref.h"
+
+#include <kernel_selector_utils.h>
+
+namespace kernel_selector {
+
+namespace {
+
+CommonDispatchData MakeDispatchData(const range_params &params) {
+  CommonDispatchData dispatchData;
```
[the remark this answers is not in the document]
Comment (serhii-pavlovskyi-altran) [reply inference-engine/thirdparty/clDNN/kernel_selector/core/actual_kernels/range/range_kernel_ref.cpp]: reformatted

File: src/plugins/intel_gpu/src/plugin/ops/range.cpp (hunk 1 above)
Comment (sshlyapn) [reply src/plugins/intel_gpu/src/plugin/ops/range.cpp]: Here I'm talking mostly about configuring gws sizes as:
`dispatchData.gws[2] = params.output.Z().v * params.output.Y().v * params.output.X().v`
And using it in kernel as:
`const uint i = get_global_id(2);`

And I want to ensure that `tensor_from_dims()` will return correct tensor with single value in correct dimension. Because I have a feeling that it will create a tensor with sizes [N, 1, 1, 1] (in bfyx format) instead of expected [1, 1, 1, N] for N requested elements - which will lead to incorrect gws sizes and results. Could you please check it?

File: inference-engine/thirdparty/clDNN/kernel_selector/core/cl_kernels/range_ref.cl (hunk 2 above)
[the remark this answers is not in the document]
Comment (sshlyapn) [reply inference-engine/thirdparty/clDNN/kernel_selector/core/cl_kernels/range_ref.cl]: I think this will be enough, thanks

File: src/plugins/intel_gpu/src/plugin/ops/range.cpp (hunk 1 above)
Comment (serhii-pavlovskyi-altran) [reply src/plugins/intel_gpu/src/plugin/ops/range.cpp]: it's not easy to check since this code isn't being run from end to end currently, but i see what you mean.
reworked

File: cmake/developer_package/add_ie_target.cmake [hunk 7]
```diff
@@ -72,7 +72,7 @@ function(addIeTarget)
     endforeach()
 
     file(GLOB_RECURSE includes ${includeSearch})
-    file(GLOB_RECURSE sources  ${sourceSearch})
+    file(GLOB_RECURSE sources CONFIGURE_DEPENDS ${sourceSearch})
```
[the remark this answers is not in the document]
Comment (sshlyapn) [reply cmake/developer_package/add_ie_target.cmake:75]: @serhii-pavlovskyi-altran, could you please revert these changes in CMake? As we discussed with @ilya-lavrenov, the advantages of this flag and how it will affect configuration time aren't clear for now

File: inference-engine/thirdparty/clDNN/kernel_selector/core/actual_kernels/range/range_kernel_ref.cpp [hunk 8]
```diff
@@ -0,0 +1,87 @@
...
+
+namespace kernel_selector {
+
+namespace {
+
+CommonDispatchData MakeDispatchData(const range_params &params) {
+    CommonDispatchData dispatchData;
+
+    dispatchData.gws = { // TODO: these could be split better
+        params.output.Batch().v, params.output.Feature().v, params.output.Z().v * params.output.Y().v
```
Comment (lznamens) [inline inference-engine/thirdparty/clDNN/kernel_selector/core/actual_kernels/range/range_kernel_ref.cpp:17]: Your current realization of this kernel doesn't support "Z" dimension, so I suppose it can be deleted here.

File: inference-engine/thirdparty/clDNN/kernel_selector/core/cl_kernels/range_ref.cl [hunk 9]
```diff
@@ -0,0 +1,11 @@
+// Copyright (C) 2021 Intel Corporation
+// SPDX-License-Identifier: Apache-2.0
+//
+
+KERNEL (range_ref)(const __global INPUT0_TYPE *startP, const __global INPUT1_TYPE *stopP,
```
Comment (lznamens) [inline inference-engine/thirdparty/clDNN/kernel_selector/core/cl_kernels/range_ref.cl:5]: Just for better understanding: is it correct that you have stopP argument but don't use it inside your kernel code?

Comment (serhii-pavlovskyi-altran) [reply inference-engine/thirdparty/clDNN/kernel_selector/core/cl_kernels/range_ref.cl:5]: stop is needed to calculate size of output vector, kernel itself has no use for it. i didn't check what it would take to stop upper layers from passing it.

File: cmake/developer_package/add_ie_target.cmake (hunk 7 above)
[the remark this answers is not in the document]
Comment (serhii-pavlovskyi-altran) [reply cmake/developer_package/add_ie_target.cmake:75]: the advage is you don't have to remember to run cmake after you've added source file and you don't have to run cmake after every branch change in case branches have different sets of files. the disadvantage is it will check globs on every build. on my system it takes about 40 milliseconds now. you can test by running `cmake -P CMakeFiles/VerifyGlobs.cmake` from build directory on this branch. it will take more time if more globs are converted

File: inference-engine/thirdparty/clDNN/kernel_selector/core/actual_kernels/range/range_kernel_ref.cpp [hunk 10]
```diff
@@ -0,0 +1,87 @@
...
+
+#include "range_kernel_ref.h"
+
+#include <kernel_selector_utils.h>
+
+namespace kernel_selector {
+
+namespace {
+
+CommonDispatchData MakeDispatchData(const range_params &params) {
```
Comment (yeonbok) [inline inference-engine/thirdparty/clDNN/kernel_selector/core/actual_kernels/range/range_kernel_ref.cpp:13]: Not an important thing: but mostly the kernels are using the convention: SetDefault for this role. Isnt it better to unify?

File: inference-engine/thirdparty/clDNN/kernel_selector/core/cl_kernels/range_ref.cl [hunk 11]
```diff
@@ -0,0 +1,11 @@
...
+                   const __global INPUT2_TYPE *stepP, __global OUTPUT_TYPE *output)
+
+{
+    const uint i = get_global_id(2);
+    output[i] = *startP + i * (OUTPUT_TYPE) * stepP;
```
Comment (yeonbok) [inline inference-engine/thirdparty/clDNN/kernel_selector/core/cl_kernels/range_ref.cl:10]: why is it multiplying OUTPUT_TYPE?

File: src/tests/functional/shared_test_classes/src/base/layer_test_utils.cpp [hunk 12]
```diff
@@ -404,7 +404,7 @@ std::vector<std::pair<ngraph::element::Type, std::vector<std::uint8_t>>> LayerTe
     auto refInputsTypes = std::vector<ngraph::element::Type>(inputs.size());
     for (std::size_t i = 0; i < inputs.size(); ++i) {
         const auto &input = inputs[i];
-        const auto &inputSize = input->byteSize();
+        const auto inputSize = input->byteSize();
```
Comment (yeonbok) [inline src/tests/functional/shared_test_classes/src/base/layer_test_utils.cpp:407]: why this is modified?

File: src/tests/functional/inference_engine/transformations/simplify_second_input_of_reshape_test.cpp [hunk 13]
```diff
@@ -18,12 +18,13 @@
 using namespace testing;
 using namespace ngraph;
 
+namespace {
```
Comment (yeonbok) [inline src/tests/functional/inference_engine/transformations/simplify_second_input_of_reshape_test.cpp:21]: why is this added?

File: inference-engine/thirdparty/clDNN/tests/test_cases/range_gpu_test.cpp [hunk 14]
```diff
@@ -0,0 +1,82 @@
...
+void doSmokeRangeAllTypes( int start, int stop, int step ) {
+    doSmokeRange<std::int8_t>(start, stop, step);
+    doSmokeRange<std::uint8_t>(start, stop, step);
+    doSmokeRange<int>(start, stop, step);
+    doSmokeRange<float>(start, stop, step);
+    doSmokeRange<std::int64_t>(start, stop, step);
+}
+
+TEST(smoke, Range) {
+    doSmokeRangeAllTypes(1, 21, 2);
```
Comment (yeonbok) [inline inference-engine/thirdparty/clDNN/tests/test_cases/range_gpu_test.cpp:79]: how about test for negative step too?

File: inference-engine/thirdparty/clDNN/kernel_selector/core/cl_kernels/range_ref.cl (hunk 11 above)
Comment (yeonbok) [reply inference-engine/thirdparty/clDNN/kernel_selector/core/cl_kernels/range_ref.cl:10]: Was it casting ? than how about using TO_OUTPUT_TYPE instead of (OUTPUT_TYPE)?

Comment (serhii-pavlovskyi-altran) [reply inference-engine/thirdparty/clDNN/kernel_selector/core/cl_kernels/range_ref.cl:10]: `OUTPUT_TYPE` is a typedef, i.e. `i` is multiplied by `*stepP` cast to `OUTPUT_TYPE`

File: src/tests/functional/shared_test_classes/src/base/layer_test_utils.cpp (hunk 12 above)
Comment (serhii-pavlovskyi-altran) [reply src/tests/functional/shared_test_classes/src/base/layer_test_utils.cpp:407]: i was confused by binding of prvalue integer to reference.
probably it's just copy-paste of line above.

File: src/tests/functional/inference_engine/transformations/simplify_second_input_of_reshape_test.cpp (hunk 13 above)
Comment (serhii-pavlovskyi-altran) [reply src/tests/functional/inference_engine/transformations/simplify_second_input_of_reshape_test.cpp:21]: as explained in description of commit 4, it fixes odr violation

File: inference-engine/thirdparty/clDNN/kernel_selector/core/actual_kernels/range/range_kernel_ref.cpp (hunk 10 above)
Comment (serhii-pavlovskyi-altran) [reply inference-engine/thirdparty/clDNN/kernel_selector/core/actual_kernels/range/range_kernel_ref.cpp:13]: renamed locally to `SetDefault`, though it's not clear what default is being set in these functions

File: inference-engine/thirdparty/clDNN/kernel_selector/core/actual_kernels/range/range_kernel_ref.cpp (hunk 8 above)
Comment (serhii-pavlovskyi-altran) [reply inference-engine/thirdparty/clDNN/kernel_selector/core/actual_kernels/range/range_kernel_ref.cpp:17]: i've dropped everything but **x**, it exposed bug in passing dimensions in my unit tests, which i've also fixed

File: inference-engine/thirdparty/clDNN/tests/test_cases/range_gpu_test.cpp (hunk 14 above)
Comment (serhii-pavlovskyi-altran) [reply inference-engine/thirdparty/clDNN/tests/test_cases/range_gpu_test.cpp:79]: done

File: cmake/developer_package/add_ie_target.cmake (hunk 7 above)
[the remark this answers is not in the document]
Comment (serhii-pavlovskyi-altran) [reply cmake/developer_package/add_ie_target.cmake]: dropped

File: inference-engine/thirdparty/clDNN/kernel_selector/core/cl_kernels/range_ref.cl (hunk 11 above)
Comment (yeonbok) [reply inference-engine/thirdparty/clDNN/kernel_selector/core/cl_kernels/range_ref.cl]: I mean, "TO_OUTPUT_TYPE" is macro for that casting.

Comment (serhii-pavlovskyi-altran) [reply inference-engine/thirdparty/clDNN/kernel_selector/core/cl_kernels/range_ref.cl]: sorry, i didn't see your second comment until page reload. converted everything with `TO_OUTPUT_TYPE`

File: inference-engine/thirdparty/clDNN/kernel_selector/core/cl_kernels/range_ref.cl (hunk 9 above)
Comment (serhii-pavlovskyi-altran) [reply inference-engine/thirdparty/clDNN/kernel_selector/core/cl_kernels/range_ref.cl]: got rid of `stopP`

**Review by sshlyapn — APPROVED** [review]
